### PR TITLE
dates: Rewrite using regular expressions

### LINKED ISSUE #893: filter: `--max-date` with year only is not inclusive
Take this example dataset:

```sh
echo -e 'strain\tdate
SEQ1\t2019-01-01
SEQ2\t2020-01-01
SEQ3\t2020-12-31
SEQ4\t2021-01-01
SEQ5\t2021-12-31
SEQ6\t2022-01-01
' > metadata.tsv
```

`--min-date` is inclusive. With `--max-date 2020`, both `2020-01-01` and `2020-12-31` pass as expected:

```sh
augur filter \
  --metadata metadata.tsv \
  --min-date 2020 \
  --output-metadata filtered.tsv
# strain	date
# SEQ3	2020-12-31
# SEQ4	2021-01-01
# SEQ2	2020-01-01
# SEQ6	2022-01-01
# SEQ5	2021-12-31
```

However, `--max-date` is not inclusive. With `--max-date 2021`, both `2021-01-01` and `2021-12-31` are expected to pass, but instead they get filtered out:

```sh
augur filter \
  --metadata metadata.tsv \
  --max-date 2021 \
  --output-metadata filtered.tsv
# strain	date
# SEQ1	2019-01-01
# SEQ3	2020-12-31
# SEQ2	2020-01-01
```

### Reason

In `--max-date 2021`, the value `2021` gets evaluated as `2021.0` by the type converter function `numeric_date`:

https://github.com/nextstrain/augur/blob/c264580153afd99ebd3816023980bdc9c040e141/augur/dates.py#L30-L32

and that value is used as `max_date` here:

https://github.com/nextstrain/augur/blob/c264580153afd99ebd3816023980bdc9c040e141/augur/filter.py#L332-L333

This means the `<= max_date` is effectively `< 2021` since the earliest ISO date `2021-01-01 ~= 2021.001`.

### Possible solution:

This has already been solved in #854. Two parts:

1. Treat `2021` as `2021-XX-XX`:

    https://github.com/nextstrain/augur/blob/110af66f18153aa9e0a05fa8484606cba8bb3aa1/augur/dates.py#L123-L133

2. Use different type converters for `--min-date` and `--max-date`, taking minimum of ambiguity for `--min-date` and maximum for `--max-date`:

    https://github.com/nextstrain/augur/blob/110af66f18153aa9e0a05fa8484606cba8bb3aa1/augur/filter.py#L22-L23

### Your environment: if running Nextstrain locally
 - Version: augur 15.0.0

### PR DESCRIPTION
**Pull request by victorlin**
### Description of proposed changes

This rewrites the date parsing functions currently used by `filter` and `frequencies` `--min-date`/`--max-date` parameters. Note that it doesn't affect the date parsing functions used for the metadata dates within `augur filter`.

### Related issue(s)

- Fixes #893
- Related to #662
- Related to #882

### Testing

- [x] Updated doctests and pytests.
- [x] Checks pass

### CODE REVIEW

File: augur/dates.py [hunk 1]
```diff
@@ -1,65 +1,259 @@
...
+# e.g. 2018-03-25
+RE_ISO_8601_DATE = re.compile(r'^\d{4}-\d{2}-\d{2}$')
+
+# complete ambiguous ISO 8601 date
+# e.g. 2018-03-XX
+RE_AMBIGUOUS_ISO_8601_DATE = re.compile(r'^[\dX]{4}-[\dX]{2}-[\dX]{2}$')
+
+# incomplete ambiguous ISO 8601 date (missing day)
+# e.g. 2018-03
+RE_AMBIGUOUS_ISO_8601_DATE_YEAR_MONTH = re.compile(r'^[\dX]{4}-[\dX]{2}$')
```
Comment (j23414) [inline augur/dates.py:37]: I guess the patterns are matching to the end of line `$` which will break on input text like "2020-10 (Day unknown)". But I assume we'll pre-process date strings to drop any parentheticals or extra spaces.

Comment (victorlin) [reply augur/dates.py:37]: Yeah, that seems beyond the scope of what's here so far, which only defines valid formats and converts to numeric.

As part of #860, we could have some logic for pre-processing date strings to these valid formats (e.g. `2020-10 (Day unknown)` -> `2020-10`).

Comment (j23414) [reply augur/dates.py:37]: Agree on beyond the scope. Makes sense to put the pre-processing in the "clean/transform/sanitize data" #860 steps

**Review by j23414 — DISMISSED** [review]
Slick, especially in the 'any_to_numeric_type_min/max'. 

As far as I can tell from reading the code, looks good to me.

File: augur/dates.py [hunk 2]
```diff
@@ -1,73 +1,271 @@
...
+# note: year-only is treated as incomplete ambiguous and must be non-negative (see RE_YEAR_ONLY)
+RE_NUMERIC_DATE = re.compile(r'^-*\d+\.\d+$')
+
+# complete ISO 8601 date
+# e.g. 2018-03-25
+RE_ISO_8601_DATE = re.compile(r'^\d{4}-\d{2}-\d{2}$')
+
+# complete ambiguous ISO 8601 date
+# e.g. 2018-03-XX
+RE_AMBIGUOUS_ISO_8601_DATE = re.compile(r'^[\dX]{4}-[\dX]{2}-[\dX]{2}$')
```
Comment (tsibley) [inline augur/dates.py:36]: Is this intended to allow values like `2022-01-X2` or `XX20-01-18`?

File: augur/dates.py [hunk 3]
```diff
@@ -1,73 +1,271 @@
...
+                day = 28
+            else:
+                day = 30
     try:
-        return float(date)
+        return date_to_numeric_capped(datetime.date(year, month, day))
     except ValueError:
-        pass
+        # catches month/day out of bounds errors
+        raise InvalidDateFormat
```
Comment (tsibley) [inline augur/dates.py:124]: Should this exception object get a similar message to the other places we throw the same type?

File: augur/dates.py [hunk 4]
```diff
@@ -1,73 +1,271 @@
...
+    2017.001
+    >>> round(date_to_numeric_capped(datetime.date(2019, 1, 1), max_numeric=2018.0), 3)
+    2018.0
+    """
+    if max_numeric is None:
+        max_numeric = date_to_numeric(datetime.date.today())
+    date_numeric = date_to_numeric(date)
+    if date_numeric > max_numeric:
+        date_numeric = max_numeric
+    return date_numeric
```
Comment (tsibley) [inline augur/dates.py:266]: This can be simplified to:

```suggestion
return min(date_numeric, max_numeric)
```

**Review by tsibley — CHANGES_REQUESTED** [review]

File: augur/dates.py (hunk 2 above)
Comment (jameshadfield) [reply augur/dates.py:36]: Can't speak for the intention here, but we shouldn't allow `X` in a date if followed by a number.

Comment (victorlin) [reply augur/dates.py:36]: Thanks for catching this. The regex really needs some work!

I'm not sure what I'm intending here and what is desired. Created a separate issue where this should be clarified before continuing work here: #989

File: augur/dates.py [hunk 5]
```diff
@@ -1,73 +1,271 @@
...
-    >>> numeric_date("2020.42")
-    2020.42
-    >>> numeric_date("2020-06-04")
-    2020.42486...
-    >>> import datetime, isodate, treetime
-    >>> numeric_date("1W") == treetime.utils.numeric_date(datetime.date.today() - isodate.parse_duration("P1W"))
-    True
+# float, negative ok
+# note: year-only is treated as incomplete ambiguous and must be non-negative (see RE_YEAR_ONLY)
+RE_NUMERIC_DATE = re.compile(r'^-*\d+\.\d+$')
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py]: Thanks for catching. I intend to allow values like `-.5` and `-2020.`. Fixed: 9e1f2309

File: augur/dates.py [hunk 6]
```diff
@@ -1,73 +1,271 @@
...
+
+# complete ISO 8601 date
+# e.g. 2018-03-25
+RE_ISO_8601_DATE = re.compile(r'^\d{4}-\d{2}-\d{2}$')
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py]: Done: 234686ec

File: augur/dates.py [hunk 7]
```diff
@@ -1,73 +1,271 @@
...
+
+# incomplete ambiguous ISO 8601 date (missing day)
+# e.g. 2018-03
+RE_AMBIGUOUS_ISO_8601_DATE_YEAR_MONTH = re.compile(r'^[\dX]{4}-[\dX]{2}$')
+
+# incomplete ambiguous ISO 8601 date (missing month and day)
+# e.g. 2018
+# and other positive ints
+# e.g. 1, 123, 12345
+RE_YEAR_ONLY = re.compile(r'^[1-9][\dX]*$')
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py]: You're right, not desirable. Fixed: c392812a

File: augur/dates.py [hunk 8]
```diff
@@ -1,73 +1,271 @@
...
+
+# also support relative dates (ISO 8601 durations) - see any_to_numeric()
+
+
+CACHE_SIZE = 8192
+# Some functions below use @lru_cache to minimize redundant operations on
+# large datasets that are likely to have multiple entries with the same date value.
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py:53]: Added a note: 226414d1

File: augur/dates.py [hunk 9]
```diff
@@ -1,73 +1,271 @@
...
+    elif ambiguity_resolver == 'max':
+        year = int(date_parts[0].replace('X', '9'))
+        month = int(date_parts[1]) if len(date_parts) > 1 and date_parts[1].isnumeric() else 12
+        if len(date_parts) == 3 and date_parts[2].isnumeric():
+            day = int(date_parts[2])
+        else:
+            if month in {1,3,5,7,8,10,12}:
+                day = 31
+            elif month == 2:
+                day = 28
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py:118]: I originally [copied this from treetime](https://github.com/neherlab/treetime/blob/92db7b9fbf2725c9065b1c8dc69b8980b096cb5a/treetime/utils.py#L380-L381) when trying to optimize for performance here. But I think caching does the job for performance now, and it should be ok to use another library that might be slightly less efficient but more concise.

File: augur/dates.py (hunk 3 above)
Comment (victorlin) [reply augur/dates.py]: Yes, done: 468f3b46

File: augur/dates.py (hunk 4 above)
Comment (victorlin) [reply augur/dates.py]: Done: dc754b7a

File: augur/dates.py [hunk 10]
```diff
@@ -1,73 +1,271 @@
...
+    >>> round(relative_iso_to_numeric('5W', from_date=datetime.date(2018, 3, 25)), 3)
+    2018.133
+    >>> round(relative_iso_to_numeric('5Y', from_date=datetime.date(2018, 3, 25)), 3)
+    2013.229
+    """
+    if from_date is None:
+        from_date = datetime.date.today()
+    if not backwards_duration_str.startswith('P'):
+        backwards_duration_str = 'P'+backwards_duration_str
+    return treetime.utils.numeric_date(from_date - isodate.parse_duration(backwards_duration_str))
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py]: Actually `date_to_numeric_capped()`, done: cb3cc177

**Comment by victorlin** [comment]
Converting to draft, at least until #989 is resolved

File: augur/dates.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsibley) [reply augur/dates.py:118]: Ah, I was suggesting something like `day = 29 if isleap(year) else 28` here.

**Comment by tsibley** [comment]
Is the plan to eventually replace the remaining treetime-based functions like `augur.date.get_numerical_dates()` with the new ones introduced here?

**Comment by victorlin** [comment]
@tsibley yes, that's my plan with #938

File: augur/dates.py [hunk 11]
```diff
@@ -43,7 +43,7 @@ class InvalidDateFormat(ValueError):
 # e.g. 2018
 # and other positive ints
 # e.g. 1, 123, 12345
-RE_YEAR_ONLY = re.compile(r'^[1-9][0-9X]*$')
+RE_YEAR_ONLY = re.compile(r'^0*[1-9][0-9X]*$')
```
Comment (tsibley) [inline augur/dates.py:46]: Just a note that I was wondering about if could simplify down to `^[0-9X]+$`, but I think I realized why not: neither `0` nor `X` alone should be allowed, and the `[1-9]` in the middle ensures this. (However, it also means that `0XXX` and `XX00` don't match.)

Comment (victorlin) [reply augur/dates.py:46]: > However, it also means that `0XXX` and `XX00` don't match.

Hmm I hadn't thought of this.

- `0XXX`: this is just the same as `XXX` which I don't think should be supported.
- `XX00`: this should be supported if `XX11` is also supported (it is currently matched). However, if in [#989](https://github.com/nextstrain/augur/issues/989) the decision is to stop supporting partial ambiguity in years, then neither `XX00` nor `XX11` should be supported.

File: augur/dates.py [hunk 12]
```diff
@@ -1,73 +1,271 @@
...
+        month = int(date_parts[1]) if len(date_parts) > 1 and date_parts[1].isnumeric() else 12
+        if len(date_parts) == 3 and date_parts[2].isnumeric():
+            day = int(date_parts[2])
+        else:
+            if month in {1,3,5,7,8,10,12}:
+                day = 31
+            elif month == 2:
+                day = 28
+            else:
+                day = 30
```
Comment (fanninpm) [inline augur/dates.py:120]: [`calendar.monthrange`](https://docs.python.org/3/library/calendar.html#calendar.monthrange) might be helpful.

**Comment by victorlin** [comment]
While re-reading through the commits, I realized it was hard to follow. Just rebased and force-pushed with more verbose commits which should hopefully help make sense of the changes here.

File: augur/dates.py (hunk 12 above)
Comment (victorlin) [reply augur/dates.py]: Thanks, using that in d6d5172a90271f1ec44472b129f1d220d0fe8125.

File: augur/dates.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/dates.py]: Using `calendar.monthrange` in d6d5172a90271f1ec44472b129f1d220d0fe8125.

File: augur/dates.py (hunk 2 above)
Comment (victorlin) [reply augur/dates.py]: The conclusion from #989 is to not support partial ambiguity here. 6a18b35c15b7892deef4a7baff3b445dab8a4a81 reflects that decision.

File: augur/dates.py (hunk 11 above)
Comment (victorlin) [reply augur/dates.py]: 6a18b35c15b7892deef4a7baff3b445dab8a4a81 removes partial ambiguity so this isn't an issue anymore.

**Comment by victorlin** [comment]
Marking as draft again since this needs to go through some more changes following merge of #1140.

**Comment by victorlin** [comment]
Closing this in favor of #1144, which is an attempt to tackle the same problems while consolidating logic simultaneously.
